src/main/java/lotto/controller/LottoController.java

```diff
@@ -38,6 +40,27 @@ private LottoPurchaseBudget getBudget() {
         }
     }
 
+    private int getManualLottoCount(LottoPurchaseBudget budget) {
+        try {
+            int manualLottoCount = view.getManualLottoCount();
+            budget.validatePurchasable(manualLottoCount);
```

Comment:
`validatePurchasable` 대신에 autoLotto 때와 동일하게 `purchaseLotto` 메서드를 호출하면 되지 않을까요? 

```suggestion
            budget.purchaseLotto(manualLottoCount);
```

Reply:
`budget.purchaseLotto(count)`메소드는 현재 buget으로 count만큼의 로또를 살수 있는지 검증한 뒤 예산을 차감하는 동작을 수행하고, 내부적으로`validatePurchasable()` 메소드를 호출합니다.

또한 `getManualLottoCount()` 메소드는 입력값이 유효한지만 판단하여 넘겨주는 작업을 수행하도록 설계했고, 실제 구매는 도메인 레이어의 `LottoTicketSeller.purchaseManualLottoTickets()`에서 `budget.purchaseLotto(count)`를 호출하는 구조입니다.

처음 계획은 리뷰어님의 말씀처럼 컨트롤러에서 구매 로직을 호출하는 것이었지만 제가 생각했을 때 다음과 같은 문제가 있어 현재와 같이 컨트롤러에서는 검증만을 수행하고 도메인 레이어에서 구매 로직을 수행하도록 구현했습니다.

- 컨트롤러에서 비용 차감 후 LottoTicketSeller에서 수동 로또를 만드는 과정에서 예외 발생시 현재로서는 롤백이 불가능 or 롤백 방안을 마련해야 함
- 로또 구입 후 비용을 차감하는 역할은 컨트롤러가 아닌 도메인 레이어의 LottoTicketSeller가 가지고 있는게 더 자연스럽다고 생각함
- `getManualLottoCount`에서는 view의 입력값을 검증하여 유효한 입력값을 전달하는 역할만 담당(유효하지 않을시 재입력) 하도록 설계

위와같은 이유로 `purchaseLotto`가 아닌 `validatePurchasable`을 호출하도록 구현했는데 리뷰어님의 의견은 어떤가요?

Reply:
`validatePurchasable`과 `purchaseLotto` 로직이 분리된다면 로또를 구매할 때마다 매번 `validatePurchasable`를 호출해줘야할거에요. 사실상 **purchaseLotto** 행위를 하려면 당연하게도 LottoPurchaseBudget 객체는 `validatePurchasable`한 상태여야만해요.  이런 기준에서 로또 구매 행위 시점에 validatePurchasable 검증하는 로직은 행위에 같이 묶여있는게 좋다고 생각합니다. (그런 의미에서 지금의 `purchaseLotto` 메서드는 좋은 것 같아요 🙂)

롤백은 개념을 고려하신 점은 매우 좋은 접근이라고생각합니다. 👍💯
저는 이 부분을 깊게 생각하지는 않았어요. 실제 스프링 프레임워크환경이라면 트랜잭션의 개념으로 묶여서 적절한 시점에 예외만 던져주면 롤백이 되기 때문이죠. 지금과 같은 콘솔 게임 정도는 예외 상황에 예외처리를하고 프로그램을 종료해버려도 괜찮다고 생각하고 그런 시나리오를 가정하고 말씀드렸어요. 😃

> 로또 구입 후 비용을 차감하는 역할은 컨트롤러가 아닌 도메인 레이어의 LottoTicketSeller가 가지고 있는게 더 자연스럽다고 생각함

이 방안은 좋아보이네요. LottoPurchaseBudget 객체혼자서 처리하지 못하는 역할을 다른 객체에게 적절히 위임해줌으로써 풀어냈다고 생각해요. 🙂


> getManualLottoCount에서는 view의 입력값을 검증하여 유효한 입력값을 전달하는 역할만 담당(유효하지 않을시 재입력) 하도록 설계

차라리 `getManualLottoCount`에서는 view에 대한 입력 값을 검증하지 않는 것이 좋다고 생각하는데, 이렇게되면 예외 발생 시, 갯수를 다시 입력받도록 예외처리하기가 쉽지 않겠네요. 🙂 지금의 문제점을 깔끔하게 해결하려면 새로운 설계가 필요할 수도 있을 것 같아요. 🤔 단지 당장 저도 떠오르는 방법이 없기에 동현님의 방식이 지금의 상황에서는 최선의 선택이지 않을까 싶어요. 🙂 (항상 설계는 트레이드 오프의 산물이니 다른 대안이 떠오르지 않는다면 지금의 방식을 취하는 것이 맞다고 생각해요)